sql/src/main/java/io/crate/operation/collect/files/CSVLineParser.java

```diff
@@ -36,51 +38,35 @@
 
 public class CSVLineParser {
 
-    private static CsvMapper csvMapper = new CsvMapper();
     private static List<Object> keyList;
     private static ObjectMapper objectMapper = new ObjectMapper();
+    private static CsvMapper csvMapper = new CsvMapper();
+    private static CsvObjectReader csvReader = csvMapper.enable(CsvParser.Feature.TRIM_SPACES)
+        .readerWithTypedSchemaFor(String.class);
 
-    public void parseHeader(String header) throws IOException {
-        keyList = csvMapper
-            .enable(CsvParser.Feature.TRIM_SPACES)
-            .readerWithSchemaFor(String.class)
-            .readValues(header.getBytes(StandardCharsets.UTF_8))
-            .readAll();
-
-        Set<Object> keySet  = new HashSet<>(keyList);
+    public void parseHeader(BufferedReader currentReader) throws IOException {
+        String header = currentReader.readLine();
+        keyList = csvReader.readValues(header.getBytes(StandardCharsets.UTF_8)).readAll();
+        Set<Object> keySet = new HashSet<>(keyList);
         if (keySet.contains("")) {
             keySet.removeAll(Collections.singleton(""));
         }
 
-        if (keySet.size() != keyList.size()) {
+        if (keySet.size() != keyList.size() || keySet.size() == 0) {
             throw new IllegalArgumentException("Invalid keys!");
         }
     }
 
     public byte[] parse(String row) throws IOException {
-        List<Object> recordList = csvMapper
-            .enable(CsvParser.Feature.TRIM_SPACES)
-            .readerWithTypedSchemaFor(String.class)
-            .readValues(row.getBytes(StandardCharsets.UTF_8))
-            .readAll();
-
-        return convertCSVToMap(recordList);
-    }
-
-    private byte[] convertCSVToMap(List<Object> recordList) throws JsonProcessingException {
-
-        if (recordList.isEmpty()) {
-            return objectMapper.writeValueAsBytes(Collections.emptyMap());
-        }
-
-        if (keyList == null || keyList.size() != recordList.size()) {
-            throw new IllegalArgumentException("Number of row entries is not equal to the number of columns");
-        }
-
+        MappingIterator<Object> iterator = csvReader.readValues(row.getBytes(StandardCharsets.UTF_8));
         HashMap<Object, Object> csvAsMap = new HashMap<>();
-        for (int i = 0; i < keyList.size(); i++) {
-            csvAsMap.put(keyList.get(i), recordList.get(i));
+        int i = 0;
+        while (iterator.hasNext()) {
+            csvAsMap.put(keyList.get(i), iterator.next());
+            i++;
```

Comment:
Shouldn't this throw an error if not all keys are filled for a row? Will we just NULL all fields which are not filled? What happens if there are more values than keys?

Reply:
At the moment, any unfilled fields would be ignored. On second thoughts, that's not the right way to handle it. I'll fix and add a test.